execute/report/report.go

```diff
@@ -118,81 +121,144 @@ func buildSingleChainReport(
+		return false, validationMetadata{}, fmt.Errorf("unable to encode report: %w", err)
+	}
+
+	maxSizeBytes := int(b.maxReportSizeBytes - b.accumulated.encodedSizeBytes)
+	if len(encoded) > maxSizeBytes {
+		b.lggr.Debugw("invalid report, report size exceeds limit", "size", len(encoded), "maxSize", maxSizeBytes)
+		return false, validationMetadata{}, nil
 	}
 
-	return finalReport, len(encoded), nil
+	return true, validationMetadata{
+		encodedSizeBytes: uint64(len(encoded)),
+	}, nil
 }
 
-// buildSingleChainReportMaxSize generates the largest report which fits into maxSizeBytes.
+// buildSingleChainReport generates the largest report which fits into maxSizeBytes.
 // See buildSingleChainReport for more details about how a report is built.
-func buildSingleChainReportMaxSize(
+func (b *execReportBuilder) buildSingleChainReport(
 	ctx context.Context,
-	lggr logger.Logger,
-	hasher cciptypes.MessageHasher,
-	tokenDataReader types.TokenDataReader,
-	encoder cciptypes.ExecutePluginCodec,
 	report plugintypes.ExecutePluginCommitDataWithMessages,
-	maxSizeBytes int,
-) (cciptypes.ExecutePluginReportSingleChain, int, plugintypes.ExecutePluginCommitDataWithMessages, error) {
-	finalReport, encodedSize, err :=
-		buildSingleChainReport(ctx, lggr, hasher, tokenDataReader, encoder, report, 0)
+) (cciptypes.ExecutePluginReportSingleChain, plugintypes.ExecutePluginCommitDataWithMessages, error) {
+	finalize := func(
+		execReport cciptypes.ExecutePluginReportSingleChain,
+		commitReport plugintypes.ExecutePluginCommitDataWithMessages,
+		meta validationMetadata,
+	) (cciptypes.ExecutePluginReportSingleChain, plugintypes.ExecutePluginCommitDataWithMessages, error) {
+		b.accumulated.encodedSizeBytes += meta.encodedSizeBytes
+		commitReport = markNewMessagesExecuted(execReport, commitReport)
+		return execReport, commitReport, nil
+	}
+	// Attempt to include all messages in the report.
+	finalReport, err :=
+		buildSingleChainReport(b.ctx, b.lggr, b.hasher, b.tokenDataReader, report, nil)
 	if err != nil {
 		return cciptypes.ExecutePluginReportSingleChain{},
-			0,
 			plugintypes.ExecutePluginCommitDataWithMessages{},
 			fmt.Errorf("unable to build a single chain report (max): %w", err)
 	}
 
-	// return fully executed report
-	if encodedSize <= maxSizeBytes {
-		report = markNewMessagesExecuted(finalReport, report)
-		return finalReport, encodedSize, report, nil
+	validReport, meta, err := b.verifyReport(ctx, finalReport)
+	if err != nil {
+		return cciptypes.ExecutePluginReportSingleChain{},
+			plugintypes.ExecutePluginCommitDataWithMessages{},
+			fmt.Errorf("unable to verify report: %w", err)
+	} else if validReport {
+		return finalize(finalReport, report, meta)
 	}
 
-	var searchErr error
-	idx := sort.Search(len(report.Messages), func(mid int) bool {
-		if searchErr != nil {
-			return false
+	finalReport = cciptypes.ExecutePluginReportSingleChain{}
+	msgs := make(map[int]struct{})
+	for i := range report.Messages {
```

Comment:
@asoliman92  this is where `sort.Search` is replaced. The algorithm now adds messages in one at a time rather than attempting a binary search. I'm still not entirely happy with either approach.